### ISSUE TITLE
Feature Request - add a way to exit an emulator and go back to the frontend

### ISSUE THREAD

**Issue by iwubcode** [mmatyas/pegasus-frontend#455]
For couch style gaming, while a few emulators have a way to map a button or set of buttons to exit, many do not.

It'd be nice if pegasus could somehow support this and have it be configurable.  Ex:  'press and hold the home button for 5 seconds to quit the current emulator and return to pegasus'.

**Comment by mmatyas** [comment]
That'd be a nice feature but I'm not sure it's technically possible on all platforms. Might work in theory for keyboards on Windows, but may not for gamepads and different OSes.

**Comment by iwubcode** [comment]
Yeah, it's not an easy task, when you take into account all the various platforms Pegasus supports.

I may be wrong about this but Pegasus already supports input handling on the various OS so it's really a question of whether you can keep using Qt for this feature.  If not, I'm sure there's some general purpose input library that could work.

For me, the launching/process management side seems more daunting.  Does pegasus quit or just hide itself when a game is launched?  The parent process (pegasus) would still have to query input while hidden.  That seems like it could be tricky.

One open-source example of a front-end doing this is attract-mode (it doesn't support Android though) where it seems you configure an 'exit button'.  I think Launchbox has support for this in some capacity too but that has even less platforms to worry about!

**Comment by mmatyas** [comment]
It's not the input handling that's problematic, but rather grabbing the input of another program or the global system. I fear there will be either security, performance or platform support issues. However if there's already a working example that this can be done, that's really promising!

**Comment by DreevCi** [comment]
Hi, I just wanted to chime in to also ask for this feature given it's been posted for a year and half. 
My use case involves Steam Link on an Nvidia Shield to stream games to the TV from my rig in another room. I've tested launching Pegasus as a Steam shortcut and it seems to work but then I cannot stop games using the Steam overlay as the task reacting to the 'Exit Game' command is Pegasus in this case. I'm trying to run emulators like Cemu or Dolphin this way but it breaks the leanback experience as I'm not able to exit the emulators without this feature. 
In LaunchBox and BigBox there is an option to exit games through a combination of keys so it must be possible to do. I tried this by running BigBox as a Steam shortcut and it works. 
As an extra feature request, in case this gets added, it be great if a sort of long press in the command can be added. In BigBox I've configured the combination of the Start and Select buttons to exit a game but the game exits immediately which seems rather harsh in case the combination was triggered by mistake. Retroarch has introduced the option to press the combination twice in order to exit (with a nice message informing of the process) so that or long press of a couple of seconds would be good.

**Comment by barolo** [comment]
If you have a dualshock you can map combo for game exiting through ds4drv. Dunno about other pads though @DreevCi

**Comment by DreevCi** [comment]
I'm using the Steam Link with its controller support so not sure if I can use any external tool. Gamepads are mapped as Xinput controllers and that's it. And I currently have 3 different types of controllers (Xbox One, PS4 and 8bitdo) so a universal way of doing this would be great. I actually saw that with Dolphin there is a hot key binding feature so in the end the emulators I'm missing are PCSX2 and (specially this one) Cemu.

**Comment by barolo** [comment]
@DreevCi with cemu you could bind xdotool  `exec sh -c 'xdotool key Escape key alt+f key e key KP_Enter'` that's what I do with ds4drv [ unmaximizes to show menu, picks File menu, picks exit, exits ] bound to a pad combo.
It actually works with any app that has a normal menu

**Comment by iwubcode** [comment]
I appreciate you weighing in @DreevCi .  I'm honestly a bit surprised others aren't running into this more often.

I  have a similar setup.  Shield device streamed from Steam on a Windows host.  At the moment, if I'm not using Dolphin, I have to walk into the other room to close the application.

**Comment by barolo** [comment]
There must be a way to bind pads under windows surely? [ I don't know honestly, I'm lucky enough to not have to use it ].
You have things like autohotkey in place of the xdootol if memory serves me right.

**Comment by DreevCi** [comment]
There are options for sure. Someone on Discord just wrote how he uses RocketLauncher + Joy2Key. I'd need to check those myself. In any case, I understand this is not a priority given the options available but it'd be great to have it as it would make the frontend experience easier for everyone, specially the not so tech savvy.

**Comment by polson** [comment]
You can do this quite nicely on Windows with [reWASD](https://www.rewasd.com/)

**Comment by ICEREG1992** [comment]
Bumping this as I just dealt with the same issue on my system. In the end, I simply set Steam Big Picture to map Start+Select on my controller to Ctrl+Q, which I then set as the Exit hotkey for all of my external emulators.

My thinking now is that, surely this can be done via Pegasus so I don't have to have Steam open in the background hogging memory? Gamepad support is already in, so while Pegasus is a background process (i.e. a game is running) it can simply be waiting for the exit hotkey (set in Controls potentially), where it'll then send keypresses to the system a la AutoHotKey, exiting the emulator. You could have the exit keypresses documented somewhere so users would know how to set their configs, OR add an optional ```exit:``` attribute for collections where a user can define what they would like to be pressed to exit the emulator.

I'm pretty new to Pegasus so forgive me for not knowing exactly what permissions that would take or whether it's actually feasible, but having this kind of feature taken care of by Pegasus for me would be the ideal, so I'm just dropping some ideas.

**Comment by iwubcode** [comment]
The program you're suggesting is a _paid_ program.  That's not really a good option for many users.  From looking around, I did find https://keysticks.net/ which is free/open-source but I'm assuming it doesn't support all the controllers Pegasus does.

This feature seems split for frontends.  I had thought this was a common feature but it seems Playnite also doesn't support it.  Meanwhile Launchbox and Attract Mode do.

I personally still run in the other room and manually quit the program.  Not sure these other programs suggested would even work for my use-case since the input is through Steamlink (app).

I'm fine with whatever mmatyas ends up deciding.  I had assumed there would be a large amount of users with similar needs but that doesn't seem to be the case.

**Comment by Nino-Bock** [comment]
I want to support this request.
For the moment I'm using RetroBat/EmulationStation which allows me to press Hotkey+Start to exit (almost) any game.
The drawback is, RetroBat is Windows only and I want a solution which works under Linux as well.
Pegasus looks really well done, however if it does not support a hotkey to kill games that is a huge downside for me.
